src/aspire/source/relion.py

```diff
@@ -227,11 +227,21 @@ def populate_metadata(self, filepath, data_folder=None, max_rows=None):
     def __str__(self):
         return f"RelionSource ({self.n} images of size {self.L}x{self.L})"
 
-    def _images(self, start=0, num=np.inf, indices=None):
-        if indices is None:
-            indices = np.arange(start, min(start + num, self.n))
-        else:
-            start = indices.min()
+    def _images(self, indices):
+        """
+        Returns particle images when accessed via the `ImageSource.images` property.
+        Loads particle images corresponding to `indices` from StarFile and .mrcs stacks.
+        :param indices: A 1-D NumPy array of integer indices.
+        :return: An `Image` object.
+        """
+
+        # check for cached images first
+        if self._cached_im is not None:
+            logger.info("Loading images from cache")
+            return self.generation_pipeline.forward(
```

Comment:
Seems like an important line to cover in tests.

Reply:
Added `cache()` tests for `RelionSource`, `CoordinateSource`, and `Simulation`